### PR TITLE
Restrict Azure NSG rules to allow external access only to load balancer IP

### PR DESCRIPTION
**Pull request by itowlson**
**What this PR does / why we need it**: On Azure, we create NSG (Network Security Group) rules on the vnet to allow external clients to access services exposed as type LoadBalancer.  At the moment, these rules have a destination of `Any`, which means that they will permit requests on the opened port to any IP within the vnet.  This PR restricts the security rules so that they admit external access only to the load balancer IP.

**Which issue this PR fixes**: None in upstream - reported as https://github.com/Azure/acs-engine/issues/1619 

**Special notes for your reviewer**: None

**Release note**: 

```release-note
Azure NSG rules for services exposed via external load balancer 
now limit the destination IP address to the relevant front end load 
balancer IP.
```

### CODE REVIEW

**Comment by colemickens** [comment]
/sig azure

cc: @jdumars @brendandburns @anhowe to possibly suggest an Azure employee to review.

If no one from Azure takes a look soon, I can do a closer review - change looks okay from a quick skim though.

**Review by seanknox — APPROVED** [review]
LGTM for me, though want to have @jackfrancis sign off too. @jdumars this is a fix to shore up expected cloud provider behavior that we'll want to land in v1.8.2 if possible.

File: pkg/cloudprovider/providers/azure/azure_test.go [hunk 1]
```diff
@@ -299,7 +299,7 @@ func TestReconcileSecurityGroupNewServiceAddsPort(t *testing.T) {
 
 	sg := getTestSecurityGroup()
 
-	sg, _, err := az.reconcileSecurityGroup(sg, testClusterName, &svc1, true)
+	sg, _, err := az.reconcileSecurityGroup(sg, testClusterName, &svc1, to.StringPtr("13.70.140.150"), true)
```
Comment (jdumars) [inline pkg/cloudprovider/providers/azure/azure_test.go]: Is this an acceptable use of a hard-coded IP?

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go [hunk 2]
```diff
@@ -257,6 +221,54 @@ func (az *Cloud) EnsureLoadBalancer(clusterName string, service *v1.Service, nod
...
+	}
+
+	az.operationPollRateLimiter.Accept()
+	glog.V(10).Infof("SecurityGroupsClient.Get(%q): start", az.SecurityGroupName)
+	sg, err := az.SecurityGroupsClient.Get(az.ResourceGroup, az.SecurityGroupName, "")
+	glog.V(10).Infof("SecurityGroupsClient.Get(%q): end", az.SecurityGroupName)
+	if err != nil {
+		return nil, err
+	}
+	sg, sgNeedsUpdate, err := az.reconcileSecurityGroup(sg, clusterName, service, lbIP, true /* wantLb */)
```
Comment (anhowe) [inline pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: Does locking down to the lbIP affect the delivery of the external IP to the node?  The LB is set to "enableFloatingIP": true, and I believe this drops the external IP address on the box.

File: pkg/cloudprovider/providers/azure/azure_test.go (hunk 1 above)
Comment (itowlson) [reply pkg/cloudprovider/providers/azure/azure_test.go]: It's in unit tests, so it is never actually used as an IP address.  But I will change it to something meaningless / invalid - the tests will still work.

Comment (jdumars) [reply pkg/cloudprovider/providers/azure/azure_test.go]: I would use an rfc 1918 address.

**Comment by lachie83** [comment]
Hey @colemickens I would like to understand if changing from ANY to the LB destination IP in the NSG rule will break anything. Can you shed some light on why it was originally set to ANY?

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 2 above)
Comment (itowlson) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: I've tested the following scenarios:

* Single node so the request is guaranteed to land on the right node
* Multiple nodes so kubeproxy has to forward the request to the right node
* Initiating a request from within the cluster to exercise the Node->LB->Node route

All of these are working, though I've only tested interactively.

Comment (itowlson) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: From our conversation I think this addresses the routing cases that you wanted to check - let me know if I'm still not covering the bases!

**Comment by colemickens** [comment]
@lachie83 Embarrassingly, I think it was just a result of not thinking about it very closely. I would feel more comfortable if it was a reference to the PIP rather than the PIP's current address, but I can't imagine any reason the PIP's IP would ever be released/renewed and be different. It seems like overall this change is safe and appropriate.

The only concern is backward compatibility, but even then I can't think of much. Just a possibility that manual action may be required to re-write the rules, for example, if the reconciliation code doesn't look at *every field* in terms of making sure all necessary rules are in place.

**Comment by lachie83** [comment]
@colemickens thanks for responding and don't be embarrassed at all. I just wanted to know if there was a reason why so that we didn't break anything. This needs some real use-case testing for both north-south (off to on-cluster) traffic paths as well as east/west given how this provider works.

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 2 above)
Comment (brendandburns) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: Can you also check to ensure that the source IP preservation still works with this:

https://kubernetes.io/docs/tutorials/services/source-ip/

Thanks!

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go [hunk 3]
```diff
@@ -257,6 +221,54 @@ func (az *Cloud) EnsureLoadBalancer(clusterName string, service *v1.Service, nod
...
+	if err != nil {
+		return nil, err
+	}
+	if sgNeedsUpdate {
+		glog.V(3).Infof("ensure(%s): sg(%s) - updating", serviceName, *sg.Name)
+		// azure-sdk-for-go introduced contraint validation which breaks the updating here if we don't set these
+		// to nil. This is a workaround until https://github.com/Azure/go-autorest/issues/112 is fixed
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: I think this has been fixed? Can you verify we're on the updated SDK and then remove this workaround?

Comment (colemickens) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: +1 (there's a number of places with this workaround, they should all be removable.)

Comment (itowlson) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: Confirmed that we are on azure-sdk-for-go 10.0.4-beta and go-autorest 8.0.0.  According to https://github.com/Azure/go-autorest/issues/112#issuecomment-278426557, the fix was merged prior to in go-autorest 8.0.0-beta.  Will remove the workaround.

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 2 above)
Comment (itowlson) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: I ran a test using the `source-ip-app` tool from the page you linked.  Without the `externalTrafficPolicy: Local` setting, the tool reported `client_address=10.244.0.1`.  With the `externalTrafficPolicy: Local` setting, the tool reported `client_address=167.220.242.83`.  So I think source IP preservation is still okay.  Thanks for flagging it!

(I'm still not sure how to automate this testing though.  It feels like our testing matrix is getting more complex by the day and ad hoc hand testing won't cut it forever...)

**Comment by rjtsdl** [comment]
I personally prefer to bind NSG rule with PIP, rather than PIP's current IP address.

**Comment by rjtsdl** [comment]
@itowlson Ivan, are we still pushing this PR? If it cannot make the timeline of 1.9, let's hold on this for now. As we have quite a big change in the azure cloud provider folder.

**Comment by kevinkim9264** [comment]
Will this patch be back-ported? or be released in next minor release by any chance?

**Comment by jdumars** [comment]
@kevinkim9264 I added the candidate label, and we can see if it gets CP'd by someone.

**Comment by kevinkim9264** [comment]
@jdumars thank you!

**Pull request merged**